ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoiceNotifier.java

```diff
@@ -136,113 +137,44 @@ private void internalTerminalBlockReached(Bytes32 executionBlockHash) {
-          });
-    }
-
-    // we have no SentForkChoiceState or doesn't build on top of the right block
-    if (parentExecutionHash.isZero()) {
-      // pre-merge
-
-      if (executionPayloadTerminalBlockHash.isEmpty()) {
-        // we are in pre-merge and terminal block is not reached, we can build a block with empty
-        // payload
-        return SafeFuture.completedFuture(Optional.empty());
-      }
-
-      if (allowPayloadIdOnTheFlyRetrieval) {
-        // we are in pre-merge and terminal block is reached
-        // try to obtain a payloadId now
-        final Bytes32 terminalBlockHash = executionPayloadTerminalBlockHash.get();
-        return requestPayloadId(terminalBlockHash, Bytes32.ZERO, parentBeaconBlockRoot);
-      }
-    }
-
-    // Merge is complete, so we must have a real payloadId, but we don't have one that matches
-
-    if (allowPayloadIdOnTheFlyRetrieval) {
-      // try to obtain a payloadId now
-      Bytes32 finalizedExecutionBlockHash;
-
-      finalizedExecutionBlockHash =
-          lastSentForkChoiceState
-              .map(ForkChoiceState::getFinalizedBlockHash)
-              .orElseGet(this::retrieveFinalizedExecutionBlockHash);
-
-      return requestPayloadId(
-          parentExecutionHash, finalizedExecutionBlockHash, parentBeaconBlockRoot);
+    final UInt64 timestamp = spec.getSlotStartTime(blockSlot, recentChainData.getGenesisTime());
+    if (forkChoiceUpdateData.isPayloadIdSuitable(parentExecutionHash, timestamp)) {
+      return forkChoiceUpdateData.getPayloadId();
+    } else if (parentExecutionHash.isZero() && !forkChoiceUpdateData.hasTerminalBlockHash()) {
+      // Pre-merge so ok to use default payload
+      return SafeFuture.completedFuture(Optional.empty());
+    } else {
+      // Request a new payload.
+
+      // to make sure that we deal with the same data when calculatePayloadAttributes asynchronously
+      // returns, we save locally the current class reference.
+      ForkChoiceUpdateData localForkChoiceUpdateData = forkChoiceUpdateData;
+      return payloadAttributesCalculator
+          .calculatePayloadAttributes(blockSlot, inSync, localForkChoiceUpdateData)
+          .thenCompose(
+              newPayloadAttributes -> {
+
+                // we make the updated local data global, reverting any potential data not yet sent
+                // to EL
+                forkChoiceUpdateData =
+                    localForkChoiceUpdateData.withPayloadAttributes(newPayloadAttributes);
+                sendForkChoiceUpdated();
+                if (forkChoiceUpdateData.getPayloadId().isCompletedNormally()
+                    && forkChoiceUpdateData.getPayloadId().join().isEmpty()) {
+                  return SafeFuture.failedFuture(
+                      new RuntimeException("Unable to obtain a payloadId"));
+                }
+                return forkChoiceUpdateData.getPayloadId();
```

Comment:
I don't think this is right - we should be performing this check regardless of whether the future is completed immediately or not.  I'd say we want something more like:

```
return forkChoiceUpdateData.getPayloadId().thenApply(payloadId -> {
    if (payloadId.isEmpty()) {
        throw new IllegalStateException("Unable to obtain a payloadId");
    }
    return payloadId;
});

Reply:
🤦